test/functional/p2p_filter.py

```diff
@@ -64,95 +92,140 @@ def set_test_params(self):
+        self.log.info("Send a tx to the node before connecting")
+        filter_address = self.nodes[0].decodescript(filter_peer.watch_script_pubkey)['addresses'][0]
+        txid = self.nodes[0].sendtoaddress(filter_address, 90)
 
-        self.log.info('Add filtered P2P connection to the node')
-        filter_node.send_and_ping(filter_node.watch_filter_init)
-        filter_address = self.nodes[0].decodescript(filter_node.watch_script_pubkey)['addresses'][0]
+        self.log.info("Send a mempool msg after connecting and check that the tx is received")
+        self.nodes[0].add_p2p_connection(filter_peer)
+        filter_peer.send_and_ping(filter_peer.watch_filter_init)
+        self.nodes[0].p2p.send_message(msg_mempool())
+        filter_peer.wait_for_tx(txid)
+        assert filter_peer.tx_received
+
+    def test_frelay_false(self, filter_peer):
+        self.log.info("Check that a node with fRelay set to false does not receive invs until the filter is set")
+        filter_address = self.nodes[0].decodescript(filter_peer.watch_script_pubkey)['addresses'][0]
+        filter_peer.tx_received = False
+        self.nodes[0].sendtoaddress(filter_address, 90)
+        # Sync to make sure the reason filter_peer doesn't receive the tx is not p2p delays
+        filter_peer.sync_with_ping()
+        assert not filter_peer.tx_received
+
+        # Clear the mempool so that this transaction does not impact subsequent tests
+        self.nodes[0].generate(1)
+
+    def test_filter(self, filter_peer):
+        filter_peer.send_and_ping(filter_peer.watch_filter_init)
+        filter_address = self.nodes[0].decodescript(filter_peer.watch_script_pubkey)['addresses'][0]
 
         self.log.info('Check that we receive merkleblock and tx if the filter matches a tx in a block')
         block_hash = self.nodes[0].generatetoaddress(1, filter_address)[0]
         txid = self.nodes[0].getblock(block_hash)['tx'][0]
-        filter_node.wait_for_merkleblock(block_hash)
-        filter_node.wait_for_tx(txid)
+        filter_peer.wait_for_merkleblock(block_hash)
+        filter_peer.wait_for_tx(txid)
 
         self.log.info('Check that we only receive a merkleblock if the filter does not match a tx in a block')
-        filter_node.tx_received = False
+        filter_peer.tx_received = False
         block_hash = self.nodes[0].generatetoaddress(1, self.nodes[0].getnewaddress())[0]
-        filter_node.wait_for_merkleblock(block_hash)
-        assert not filter_node.tx_received
+        filter_peer.wait_for_merkleblock(block_hash)
+        assert not filter_peer.tx_received
 
         self.log.info('Check that we not receive a tx if the filter does not match a mempool tx')
-        filter_node.merkleblock_received = False
-        filter_node.tx_received = False
+        filter_peer.merkleblock_received = False
+        filter_peer.tx_received = False
         self.nodes[0].sendtoaddress(self.nodes[0].getnewaddress(), 90)
-        filter_node.sync_with_ping()
```

Comment:
Weird that there are two `sync_with_ping`s here. @MarcoFalke would it be correct for me to remove one (since I'm touching the lines anyway)? From what I understand, one `sync_with_ping` is enough to ensure that the messages were sent and processed, and we'll be grabbing `mininode_lock` when we check the `*_received` values.

Reply:
There are no messages sent at all, so using `sync_with_ping` is odd, but I didn't know what else to use.

* The first sync_with_ping is needed for any imaginary invs on the wire (from `SendMessages`) to be flushed to the filter peer
* The seconds one is needed for our imaginary getdata (in reply to the imaginary inv) to be flushed to the node (and to be processed)

Reply:
Ohh I see what you mean. If only 1 is sent, the order could be:

1. send `ping`
2. receive `inv`, send `getdata`
3. receive `pong`, can move on
5. assert not received
6. receive data (too late)

With 2, it has to be:
1. send `ping1`
2. receive `inv`, send `getdata`
3. receive `pong1`, send `ping2`
4. any response to `getdata` must be received before `pong2`
5. assert

Is this correct?

Reply:
Jup, that's the idea